File: core/src/main/kotlin/dk/lifelist/core/Checklist.kt
package dk.lifelist.core

/**
 * Everything Denmark has, against everything you have found.
 *
 * The app's taxonomy is its *output space* — the species a model was trained on — and using it
 * as the denominator was fine while the app only answered questions. The moment a screen offers
 * to show what you have **not** found, the missing species have to be missing from Denmark's
 * list rather than from ours, or the app quietly reports a weevil family complete at 64 of 490
 * (VERIFICATION §81).
 *
 * So this is a second, larger tree that the model knows nothing about: 26,722 species in 2,461
 * families, from GBIF's Danish occurrence records. Most of it the camera cannot identify, and
 * that is not a defect to hide — it is the difference between a list of what a model can do and
 * a list of what is out there.
 */
data class ChecklistSpecies(
    val taxonId: Int,
    val scientificName: String,
    val vernacularEn: String?,
    val familyId: Int?,
    /** Danish occurrence records. Recording effort, not abundance (§72). */
    val records: Int,
    /** Whether the bundled model can identify this one, or it can only be ticked by hand. */
    val identifiable: Boolean,
) {
    val name: String get() = vernacularEn ?: scientificName
}

data class ChecklistFamily(
    val taxonId: Int,
    val scientificName: String,
    val vernacularEn: String?,
    /** Ancestor keys, coarsest first. The group is read off this with [LifeList.DEFAULT_GROUPS]. */
    val lineage: List<Int>,
    /** Species of this family recorded in Denmark. */
    val species: Int,
    /** How many of those the model can identify. */
    val identifiable: Int,
) {
    val name: String get() = vernacularEn ?: scientificName
}

class Checklist(
    val families: Map<Int, ChecklistFamily>,
    val species: Map<Int, ChecklistSpecies>,
) {
    /** Species of a family, unordered. Empty for a family this checklist has never heard of. */
    private val byFamily: Map<Int, List<ChecklistSpecies>> =
        species.values.filter { it.familyId != null }.groupBy { it.familyId!! }

    fun membersOf(familyId: Int): List<ChecklistSpecies> = byFamily[familyId].orEmpty()

    /**
     * The group a family belongs to, from its own lineage.
     *
     * Read at the family rather than the species, so a species' group is a fact about its
     * family and cannot disagree with its neighbours'.
     */
    fun groupOf(family: ChecklistFamily, groups: List<Group> = DEFAULT_GROUPS): String =
        groups.firstOrNull { it.taxonId in family.lineage }?.label ?: UNGROUPED

    fun familiesIn(group: String, groups: List<Group> = DEFAULT_GROUPS): List<ChecklistFamily> =
        families.values.filter { groupOf(it, groups) == group }

    /** A name for a taxon this checklist knows and the model does not. */
    fun nameOf(taxonId: Int): String? = species[taxonId]?.name ?: families[taxonId]?.name

    companion object {
        val EMPTY = Checklist(emptyMap(), emptyMap())
    }
}

/**
 * The index: what you have found, laid over what there is.
 *
 * Pure, so the arithmetic that decides "3 of 74" can be tested without an asset or a phone.
 */
object Index {

    data class FamilyLine(
        val family: ChecklistFamily,
        /** Distinct species of this family on your list. */
        val found: Int,
    ) {
        val total: Int get() = family.species
        val fraction: Float get() = if (total <= 0) 0f else found.toFloat() / total
        val complete: Boolean get() = total > 0 && found >= total

        /** Species here the camera cannot name, so the only way on to the list is by hand. */
        val byHandOnly: Int get() = (total - family.identifiable).coerceAtLeast(0)
    }

    data class Member(val species: ChecklistSpecies, val seen: Boolean)

    /**
     * Which species are on the list, as taxon ids.
     *
     * A record kept at genus or family is deliberately **not** counted here. "How many of the
     * 74 Danish ducks have you found" is a question about species, and a record that says
     * *Anas* is an honest record and not one of the seventy-four (§19).
     */
    fun found(records: List<Record>): Set<Int> = records.map { it.taxonId }.toSet()

    /**
     * A group's families, fullest first.
     *
     * Size rather than alphabet, and deliberately not "the ones you have started" — an order
     * that changes as you play makes the page you learned unlearnable. Ducks, waders and gulls
     * at the top of the birds is a better first screen than Aegithalidae.
     */
    fun families(
        checklist: Checklist,
        group: String,
        records: List<Record>,
        groups: List<Group> = DEFAULT_GROUPS,
    ): List<FamilyLine> {
        val seen = found(records)
        return checklist.familiesIn(group, groups)
            .map { family ->
                FamilyLine(
                    family = family,
                    found = checklist.membersOf(family.taxonId).count { it.taxonId in seen },
                )
            }
            .sortedWith(compareByDescending<FamilyLine> { it.total }.thenBy { it.family.name })
    }

    /**
     * A family's species: the ones you have first, then the rest commonest first.
     *
     * Commonest, not alphabetical. The roster is a to-do list, and the useful order for a
     * to-do list of species is the one that puts what you might actually meet at the top —
     * which is what the Danish record count is for. Alphabetical would open the ducks on a
     * vagrant nobody has seen since 1987.
     */
    fun members(
        checklist: Checklist,
        familyId: Int,
        records: List<Record>,
    ): List<Member> {
        val seen = found(records)
        return checklist.membersOf(familyId)
            .map { Member(it, it.taxonId in seen) }
            .sortedWith(
                compareByDescending<Member> { it.seen }
                    .thenByDescending { it.species.records }
                    .thenBy { it.species.scientificName }
            )
    }

    /** One group's standing: species found, species Denmark has. */
    data class GroupLine(val label: String, val found: Int, val total: Int, val families: Int) {
        val fraction: Float get() = if (total <= 0) 0f else found.toFloat() / total
    }

    fun groups(
        checklist: Checklist,
        records: List<Record>,
        groups: List<Group> = DEFAULT_GROUPS,
    ): List<GroupLine> {
        val seen = found(records)
        val labels = (groups.map { it.label } + UNGROUPED).distinct()
        val byGroup = checklist.families.values.groupBy { checklist.groupOf(it, groups) }
        return labels.map { label ->
            val families = byGroup[label].orEmpty()
            GroupLine(
                label = label,
                found = families.sumOf { family ->
                    checklist.membersOf(family.taxonId).count { it.taxonId in seen }
                },
                total = families.sumOf { it.species },
                families = families.size,
            )
        }.filter { it.total > 0 }
    }
}

File: training/src/lifelist_train/checklist.py
"""Every species recorded in Denmark, not only the ones the model can name.

The app's taxonomy is its *output space*: 3,482 species that cleared twenty iNaturalist
photographs, which is a fact about how often people photograph a thing rather than about what
lives here. That was fine while the app only ever answered questions. It stops being fine the
moment a screen offers to show you what you have **not** found, because then the missing
species have to be missing *from Denmark's list*, not from ours.

The gap is not small. Against the Danish Red List's own family totals the model can name a
median 40% of a family: 64 of 490 weevils, 2 of 228 webcaps. An index built on the model's
vocabulary would show a weevil family as complete at 64 and never mention the other 426.

So the denominator is measured here instead. GBIF's Danish occurrence facets give every species
key ever recorded in the country with a count; above a threshold of independent records, that is
a checklist. **The list you can scroll is the denominator** — a count from one source over a
list from another produces "3 of 490" above 380 rows, and an index whose arithmetic disagrees
with its own contents is worse than one that admits a smaller world.
"""

from __future__ import annotations

from collections.abc import Iterable, Mapping, Sequence
from dataclasses import dataclass

#: Ranks that name something findable. `subspecies` is deliberately absent: a life list ticks a
#: species, and two subspecies of one bird are one tick.
LEAF_RANKS = frozenset({"species"})

#: GBIF's own word for "we are not sure this is a real taxon". Kept out rather than shown.
ACCEPTED = "ACCEPTED"

#: The kingdoms a life list is of. GBIF keys.
#:
#: Denmark's occurrence records include 1,425 bacteria, 910 chromists — mostly diatoms — and
#: 149 protozoa above five records. They are real records of real organisms and they do not
#: belong on a list of things you went out and found: nobody ticks a *Bacillus*, and a diatom
#: is not identifiable from a phone. Left in, they would be 2,500 species of unreachable
#: denominator, quietly making every group look less complete than it is.
LIVING_KINGDOMS: frozenset[int] = frozenset({
    1,  # Animalia
    5,  # Fungi
    6,  # Plantae
})


@dataclass(frozen=True, slots=True)
class Species:
    """One line of the checklist."""

    key: int
    scientific_name: str
    family: int | None
    records: int
    """Danish occurrence records. Recording effort, not abundance — see §72."""

    vernacular_en: str | None = None
    vernacular_da: str | None = None

    @property
    def placed(self) -> bool:
        return self.family is not None


@dataclass(frozen=True, slots=True)
class Family:
    """A family, and the lineage the app needs to put it in a group."""

    key: int
    scientific_name: str
    lineage: tuple[int, ...]
    """Ancestor keys, kingdom first. The app tests its own group keys against this, so the
    grouping has exactly one implementation and it is the one already shipped."""

    vernacular_en: str | None = None


def wanted(
    record: Mapping[str, object],
    min_records: int,
    kingdoms: frozenset[int] = LIVING_KINGDOMS,
) -> bool:
    """Does this fetched record belong on the checklist?

    Four filters, each for its own reason. `DOUBTFUL` is GBIF saying it does not believe the
    taxon; a rank other than species is a genus or a subspecies, neither of which is a tick;
    the record threshold is what separates "occurs in Denmark" from "somebody once typed this
    name into a database"; and the kingdom is what separates a life list from a census.
    """
    if record.get("status") != ACCEPTED:
        return False
    if record.get("rank") not in LEAF_RANKS:
        return False
    if int(record.get("records") or 0) < min_records:
        return False
    lineage = record.get("lineage") or {}
    kingdom = lineage.get("kingdom") if isinstance(lineage, Mapping) else None
    return kingdom is not None and int(kingdom) in kingdoms


def family_key(record: Mapping[str, object]) -> int | None:
    """The family a record sits in, or None.

    Genuinely absent for a couple of hundred species — some algae and fungi are placed no
    finer than an order — and those are counted and reported rather than quietly dropped.
    """
    lineage = record.get("lineage") or {}
    key = lineage.get("family") if isinstance(lineage, Mapping) else None
    return int(key) if key is not None else None


def lineage_of(record: Mapping[str, object]) -> tuple[int, ...]:
    """Ancestor keys, coarsest first, for the grouping the app already does."""
    lineage = record.get("lineage") or {}
    if not isinstance(lineage, Mapping):
        return ()
    order = ("kingdom", "phylum", "class", "order", "family", "genus")
    return tuple(int(lineage[rank]) for rank in order if lineage.get(rank) is not None)


def families(records: Iterable[Mapping[str, object]]) -> dict[int, Family]:
    """Every family the kept records mention, with the lineage they share."""
    out: dict[int, Family] = {}
    for record in records:
        key = family_key(record)
        if key is None or key in out:
            continue
        names = record.get("lineage_names") or {}
        name = names.get("family") if isinstance(names, Mapping) else None
        # Cut the lineage at the family: a family's group cannot depend on which of its
        # species happened to be seen first.
        lineage = lineage_of(record)
        if key in lineage:
            lineage = lineage[: lineage.index(key) + 1]
        out[key] = Family(key=key, scientific_name=str(name or key), lineage=lineage)
    return out


def counts(
    species: Sequence[Species],
    identifiable: frozenset[int] | set[int],
) -> dict[int, tuple[int, int]]:
    """Per family: how many species Denmark has, and how many the model can name.

    Both numbers ship. The first is the denominator the index counts against; the second is
    what the app says about itself, and keeping them apart is the whole point — "3 of 64 named,
    490 in Denmark" says two true things where one number could only say a false one.
    """
    out: dict[int, list[int]] = {}
    for one in species:
        if one.family is None:
            continue
        tally = out.setdefault(one.family, [0, 0])
        tally[0] += 1
        if one.key in identifiable:
            tally[1] += 1
    return {key: (total, known) for key, (total, known) in out.items()}


def document(
    species: Sequence[Species],
    family_index: Mapping[int, Family],
    identifiable: frozenset[int] | set[int],
    *,
    country: str,
    min_records: int,
    fetched: str,
) -> dict:
    """The shipped asset.

    Keyed by taxon id throughout, because that is what a record stores and what the app looks
    up. Families carry the lineage; species carry only their family, so a species' group is a
    fact about its family and cannot disagree with its neighbours'.
    """
    per_family = counts(species, identifiable)
    return {
        "country": country,
        "source": "GBIF occurrence facets and backbone, https://www.gbif.org",
        "min_records": min_records,
        "fetched": fetched,
        "note": [
            "Species recorded in this country at least min_records times. The denominator ",
            "for 'what you have not found yet' — see VERIFICATION.md section 81. `model` ",
            "marks the ones the bundled model can identify; the rest are ticked by hand.",
        ],
        "families": {
            str(key): {
                "name": family.scientific_name,
                "vernacular_en": family.vernacular_en,
                "lineage": list(family.lineage),
                "species": per_family.get(key, (0, 0))[0],
                "model": per_family.get(key, (0, 0))[1],
            }
            for key, family in sorted(family_index.items())
            if per_family.get(key, (0, 0))[0] > 0
        },
        "species": {
            str(one.key): {
                "name": one.scientific_name,
                "vernacular_en": one.vernacular_en,
                "family": one.family,
                "records": one.records,
                "model": one.key in identifiable,
            }
            for one in sorted(species, key=lambda s: s.scientific_name)
        },
    }
